NAS-102815 / 11.3 / Rollback specific ipv6 changes

sonicaj opened the pull request:
This commit rollbacks some specific ipv6 changes while retrieving default gateway as they introduce regressions to ipv6 operations while retrieving default gateway.

sonicaj committed:
Rollback specific ipv6 changes

This commit rollbacks some specific ipv6 changes while retrieving default gateway as they introduce regressions to ipv6 operations while retrieving default gateway.

sonicaj commented:
@dlangille and @joachimmathes can you please test these changes for me and make sure they support your use cases as desired ?
Thank you!

dlangille commented:
This will be my first task this morning.  I should have feedback in about 90 minutes.

dlangille commented:
That was sooner than I thought. I won't be able to work on this for another hour or so.  But there are pkg-plist errors:

```
===> Parsing plist
===> Checking for items in STAGEDIR missing from pkg-plist
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/PKG-INFO
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/SOURCES.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/dependency_links.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/entry_points.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/requires.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/top_level.txt
===> Checking for items in pkg-plist which are not in STAGEDIR
===> Error: Plist issues found.
*** Error code 1
```

I run iocage only from packages. The package is failing to build.

dlangille commented:
Also, I am applying the patch against master as found with commit d298ad4

sonicaj commented:
@dlangille can you please try building from git directly ( using Makefile ? ). Going through the ports issue first would mean having us take a detour for now and going through the ports one first.
You should uninstall first using Makefile and then install.

dlangille commented:
Yes, I can, but I will need help with pulling in the merge request.

dlangille commented:
Oh wait, or do I just git clone from https://github.com/iocage/iocage/tree/NAS-102815 ?

sonicaj commented:
> Oh wait, or do I just git clone from https://github.com/iocage/iocage/tree/NAS-102815 ?

yes please

dlangille commented:
See https://gist.github.com/dlangille/db16d15f213ef9e262b61cf82f257272

dlangille commented:
```
[dan@knew:~/src/iocage] $ sudo ./iocage start toiler
No bridge for interface ix0found in configuration.
[dan@knew:~/src/iocage] $ 
```

To be fair, I do not know if this is iocage or my configuration.

sonicaj commented:
I think you didn't checkout to this branch @dlangill, can you please confirm ?

dlangille commented:
Well:

```
[dan@knew:~/src/iocage] $ git branch
* master
[dan@knew:~/src/iocage] $ 

```

dlangille commented:
I see what I did.  I went to https://github.com/iocage/iocage/tree/NAS-102815 and clicked on Clone or Download, and it gave me git@github.com:iocage/iocage.git

I blindly used it assuming it was the right URL.

dlangille commented:
progress: https://gist.github.com/dlangille/db16d15f213ef9e262b61cf82f257272#file-7-starting-the-jail

```
[dan@knew:~/src/iocage] $ sudo ./iocage start toiler
* Starting toiler
  + Started OK
  + Using devfs_ruleset: 34
  + Configuring VNET FAILED
  route: fe80::20d:b9ff:fe33:8716%ix0: hostname nor servname provided, or not known

Stopped toiler due to VNET failure
[dan@knew:~/src/iocage] $ 

```

sonicaj commented:
Perfect. Okay can you please share the output of `ifconfig` and `netstat -nr` ? This issue wasn't faced earlier because ipv6 default route wasn't set explicitly each time we started a jail with ipv6 capabilities ( one of the changes introduced in the original @joachimmathes PR )

sonicaj commented:
@dlangille you can mask the ip addresses etc ( for privacy concerns, i would like to have a peek at your network setup - it seems we do need to replace the part after `%` from default route, but let's figure out the how part ).

dlangille commented:
Added ifconfig and netstat -nr to the gist: https://gist.github.com/dlangille/db16d15f213ef9e262b61cf82f257272

sonicaj commented:
@dlangille if you run the following command and then start the same jail, does it start up correctly ?
```
iocage set defaultrouter6="fe80::20d:b9ff:fe33:8716%epair0b" jail_name
```

dlangille commented:
Yes!  I will wait for monitoring to finish verification.

```
[dan@knew:~/src/iocage] $ sudo ./iocage set defaultrouter6="fe80::20d:b9ff:fe33:8716%epair0b" toiler 
defaultrouter6: auto -> fe80::20d:b9ff:fe33:8716%epair0b

[dan@knew:~/src/iocage] $ sudo ./iocage start toilerr
* Starting toiler
  + Started OK
  + Using devfs_ruleset: 34
  + Configuring VNET OK
  + Using IP options: vnet
  + Starting services OK
  + Executing poststart OK
[dan@knew:~/src/iocage] $ 
```

dlangille commented:
All green now on Nagios. :)

sonicaj commented:
Good to hear, i'll sort this out to see how best to make the replacement. Once done, please stay with me for a while and we can test this PR once more to be certain we are not missing something essential after my changes.

dlangille commented:
I am here for at least 7 more hours. If you are on irc, or some instant messaging, I can attend to any fixes sooner.

joachimmathes commented:
As I expected this change does not work for us, since the link local address won't be converted to `epair0b` for the jail, if `"defaultrouter6": "auto"`:
```
iocage start vprot01
* Starting vprot01
  + Started OK
  + Using devfs_ruleset: 23
  + Configuring VNET FAILED
  route: fe80::8%inet0: hostname nor servname provided, or not known
```

sonicaj commented:
Yup @joachimmathes, i am working on correcting that ;)
Should be up in a few and then you and @dlangille can test to make sure it doesn't regress

joachimmathes commented:
Hmm, that was basically what I had implemented in `get_default_gateway()`.

sonicaj committed:
Correctly set default route for ipv6

This commit correctly sets default route for ipv6 ensuring that we have the right default gateway identified for the jail when setting ipv6 default route.

dlangille commented:
I tried the most recent commit (git pull, make, stop, set defaultroute6=auto, start): issue persisted.

sonicaj commented:
It hasn't been corrected in that. It addresses a differernt complex issue. I'll update you in a few as soon as it's live. Please try it then :)

sonicaj committed:
Improve gateway warning

This commit adds address family to gateway warning to make it more understandable.

sonicaj commented:
@dlangille, @joachimmathes can you please test this to ensure that it suits your use cases as desired ? Thank you

dlangille commented:
starting.

dlangille commented:
Success here with restating that jail with auto

```
[dan@knew:~/src/iocage] $ ./iocage get all toiler | grep default 
defaultrouter:10.55.0.1
defaultrouter6:auto
vnet_default_interface:auto
[dan@knew:~/src/iocage] $ 
```

sonicaj commented:
That is good news, @dlangille can you please make sure that this hasn't regressed anywhere else for you like ipv4 etc ? ( it shouldn't, but better safe then sorry ).
Let's wait for @joachimmathes now.

dlangille commented:
Part of the output for iocage list -l:

```
+-----+--------------+------+-------+------+-----------------+---------------------------------------------------+--------------------------------------------+----------+----------+
| 22  | toiler       | off  | up    | jail | 12.0-RELEASE-p9 | vnet0|10.55.0.13/24                               | vnet0|[redacted             ]:7e45:0:53/64 | -        | no       |
+-----+--------------+------+-------+------+-----------------+---------------------------------------------------+--------------------------------------------+----------+----------+
```

Both IP4 and IPv6 here. I will reboot this server..... oh that won't help, iocage is not in the expected location.

sonicaj commented:
@dlangille it should work as expected when you reboot the server

dlangille commented:
Yes, but the rc.d script is not in place: /usr/local/etc/rc.d/iocage

I will build a new package when this hits master. What is the ETA on that?

sonicaj commented:
As soon as @joachimmathes confirms ipv6 is good for his use case, then.

sonicaj commented:
> Hmm, that was basically what I had implemented in `get_default_gateway()`.

@joachimmathes yes indeed it was. However that required changes to `interfaces` property before a jail could be started to ensure that ipv6 default route was set up properly and we lost the ability to have it working out of the box. So this PR improves that bit.

sonicaj committed:
Update tests

This commit updates tests for retrieving default gateway.

joachimmathes commented:
Testing is in progress...

joachimmathes commented:
@sonicaj Test was successful! :+1:

dlangille commented:
I may be jumping on this prematurely, but a build on master now yields these issues:

```
===> Checking for items in STAGEDIR missing from pkg-plist
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/PKG-INFO
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/SOURCES.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/dependency_links.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/entry_points.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/requires.txt
Error: Orphaned: %%PYTHON_SITELIBDIR%%/iocage_lib-1.2rc0-py%%PYTHON_VER%%.egg-info/top_level.txt
===> Checking for items in pkg-plist which are not in STAGEDIR

```

dlangille commented:
Ahh, I see what is going on.

`poudriere testport` tosses out those errors.

`poudriere bulk` succeeds.

dlangille commented:
```
[dan@knew:~/src/iocage] $ sudo pkg upgrade py36-iocage-devel
Updating local repository catalogue...
Fetching meta.txz: 100%    820 B   0.8kB/s    00:01    
Fetching packagesite.txz: 100%  228 KiB 233.8kB/s    00:01    
Processing entries: 100%
local repository update completed. 883 packages processed.
All repositories are up to date.
The following 1 package(s) will be affected (of 0 checked):

Installed packages to be UPGRADED:
	py36-iocage-devel: 1.0.0.20190801,1 -> 1.0.0.20190809,1

Number of packages to be upgraded: 1

232 KiB to be downloaded.

Proceed with this action? [y/N]: y
[1/1] Fetching py36-iocage-devel-1.0.0.20190809,1.txz: 100%  232 KiB 238.0kB/s    00:01    
Checking integrity... done (0 conflicting)
[1/1] Upgrading py36-iocage-devel from 1.0.0.20190801,1 to 1.0.0.20190809,1...
[1/1] Extracting py36-iocage-devel-1.0.0.20190809,1: 100%
[dan@knew:~/src/iocage] $ ': sudo iocage restart toiler
* Stopping toiler
  + Executing prestop OK
  + Stopping services OK
  + Tearing down VNET OK
  + Removing devfs_ruleset: 35 OK
  + Removing jail process OK
  + Executing poststop OK
* Starting toiler
  + Started OK
  + Using devfs_ruleset: 35
  + Configuring VNET OK
  + Using IP options: vnet
  + Starting services OK
  + Executing poststart OK
[dan@knew:~/src/iocage] $ 

```

dlangille commented:
Thank you for your help.  I appreciate it.

dlangille commented:
FYI, rebooted the server to test that.  All green on Nagios. All 17 jails came up, including the one which had `vnet` issues. Thanks again.
